### ISSUE TITLE
[Question] How is this library compared to huggingface's tokenizer?

### ISSUE THREAD

**Issue by liebkne** [guillaume-be/rust-tokenizers#49]
Hi, thanks for the awesome libraries. I'm new to rust and the community is enthusiastic.  I have some questions about the tokenizer library. How is rust-tokenizers compared to huggingface's tokenizer? (Performance, tokenized result etc.) Can I use them interchangably with rust-bert?

**Comment by guillaume-be** [comment]
Hello @liebkne ,

Thank you for the interest in this library. This library is very similar to Hugging Face's tokenizers. As a matter of fact, this was developed as a port of the Python tokenizers in the Transformers library in October-November 2019. As far as I know, the development of the Rust version of Tokenizers was initiated around the same time frame and released in January 2020. I kept developing this library and use it as a default since both Hugging Face's Tokenizers and Rust-bert were in active development in 2020, and developing against two moving targets would have been challenging.

The direct consequence is that the tokenization results are expected to be identical. In fact, the integration tests for this crate compare the tokenization results on large datasets and on edge cases (accents and special characters) between this library and the tokenizers accessed via the Transformers library. 

I unfortunately do not have benchmarks available to the Rust implementation of Hugging Face's tokenizers - I ran some comparison to the Python version end of 2019 (showing as expected a significant advantage for the Rust version).

`rust-bert` has several levels of API:
- at a low level, it exposes implementations of state-of-the-art transformers models. These require a tensor input and tensor output. The user can prepare her/him-self these tensors, using either this library for tokenization or Hugging Face's tokenizers. Since the tokenization result is expected to be identical they can both be used for this purpose. I invite you to check [sbert](https://github.com/cpcdoy/rust-sbert) where the author built a sentence encoding library where both tokenization libraries can be used.
- at a higher level, Rust-bert exposes ready-to-use pipelines that can be set up in a few lines of code. For these, the tokenization comes from the present library and no options for using Hugging face's tokenizers is available yet.

I hope this helps.
